src/ir/module-splitting.cpp

```diff
@@ -670,42 +716,54 @@ void ModuleSplitter::indirectCallsToSecondaryFunctions() {
     }
   };
-  CallIndirector(*this).walkModule(&primary);
+  CallIndirector callIndirector(*this);
+  callIndirector.walkModule(&primary);
+  for (auto& [mod, secondaryPtr] : secondaryPtrMap) {
+    callIndirector.walkModule(secondaryPtr.get());
+  }
 }
 
 void ModuleSplitter::exportImportCalledPrimaryFunctions() {
   // Find primary functions called/referred in the secondary module.
-  ModuleUtils::ParallelFunctionAnalysis<std::vector<Name>> callCollector(
-    secondary, [&](Function* func, std::vector<Name>& calledPrimaryFuncs) {
-      struct CallCollector : PostWalker<CallCollector> {
-        const std::set<Name>& primaryFuncs;
-        std::vector<Name>& calledPrimaryFuncs;
-        CallCollector(const std::set<Name>& primaryFuncs,
-                      std::vector<Name>& calledPrimaryFuncs)
-          : primaryFuncs(primaryFuncs), calledPrimaryFuncs(calledPrimaryFuncs) {
-        }
-        void visitCall(Call* curr) {
-          if (primaryFuncs.count(curr->target)) {
-            calledPrimaryFuncs.push_back(curr->target);
+  //   using RefFuncMap = InsertOrderedMap<Name, std::vector<RefFunc*>>;
+  using CalledPrimaryToModules = std::map<Name, std::set<Module*>>;
+  for (auto& [mod, secondaryPtr] : secondaryPtrMap) {
+    Module* secondary = secondaryPtr.get();
+    ModuleUtils::ParallelFunctionAnalysis<CalledPrimaryToModules> callCollector(
+      *secondary,
+      [&](Function* func, CalledPrimaryToModules& calledPrimaryFuncs) {
+        struct CallCollector : PostWalker<CallCollector> {
+          const std::unordered_set<Name>& primaryFuncs;
+          CalledPrimaryToModules& calledPrimaryFuncs;
+          CallCollector(const std::unordered_set<Name>& primaryFuncs,
+                        CalledPrimaryToModules& calledPrimaryFuncs)
+            : primaryFuncs(primaryFuncs),
+              calledPrimaryFuncs(calledPrimaryFuncs) {}
+          void visitCall(Call* curr) {
+            if (primaryFuncs.count(curr->target)) {
+              calledPrimaryFuncs[curr->target].insert(getModule());
+            }
           }
-        }
-        void visitRefFunc(RefFunc* curr) {
-          if (primaryFuncs.count(curr->func)) {
-            calledPrimaryFuncs.push_back(curr->func);
+          void visitRefFunc(RefFunc* curr) {
+            if (primaryFuncs.count(curr->func)) {
+              calledPrimaryFuncs[curr->func].insert(getModule());
+            }
           }
-        }
-      };
-      CallCollector(primaryFuncs, calledPrimaryFuncs).walkFunction(func);
-    });
-  std::set<Name> calledPrimaryFuncs;
-  for (auto& entry : callCollector.map) {
-    auto& calledFuncs = entry.second;
-    calledPrimaryFuncs.insert(calledFuncs.begin(), calledFuncs.end());
-  }
+        };
+        CallCollector collector(primaryFuncs, calledPrimaryFuncs);
+        collector.setModule(secondary);
+        collector.walkFunction(func);
+      });
+
+    CalledPrimaryToModules calledPrimaryToModules;
```

Comment:
I wonder if we need to do a `walkModuleCode` or similar here to pick up function references in globals, too. That can be investigated separately, though.